crates/ty_python_semantic/src/types/infer/builder.rs

```diff
@@ -9274,6 +9432,22 @@ impl GenericContextError {
     }
 }
 
+/// Dictates the behavior when an expression is inferred multiple times.
+#[derive(Default, Debug, Clone, Copy)]
+enum MultiInferenceState {
+    /// Panic if the expression has already been inferred.
+    #[default]
+    Panic,
+
+    /// Store the intersection of all types inferred for the expression.
+    Intersect {
+        // Determines whether or not a given expression is required to be assignable to its type context
+        // despite it being inferred multiple times, i.e. whether eager diagnostics are appropriate, or a
+        // fallback type should be assumed.
+        fallback: bool,
```

Comment:
The reason we even eagerly report diagnostics during the inference of `TypedDict` literals in the first place, is because a later assignability diagnostic would not have context for missing or invalid key diagnostics. Otherwise, I don't think we should be emitting diagnostics based on the type context, which acts more as a type hint.

If we had an internal implicit `TypedDict` representation we may be able to avoid this, though we also wouldn't even need bidirectional type inference for `TypedDict` literals in that case.